source/grid/reference_cell.cc

```diff
@@ -807,6 +807,256 @@ ReferenceCell::gmsh_element_type() const
+        shifted_vertices[i][d] += 2.0;
+    Point<3> shifted_p = p;
+    for (unsigned int d = 0; d < 3; ++d)
+      shifted_p[d] += 2.0;
+
+    // We are projecting onto a face of a unit cell so the vectors describing
+    // its local coordinate system should be orthogonal. We don't know which of
+    // these three vectors are mutually orthogonal for triangles so that
+    // requires an extra check
+    Tensor<1, 3>   e0;
+    Tensor<1, 3>   e1;
+    const Point<3> vertex = shifted_vertices[0];
+    // Triangles are difficult because of two cases:
+    // 1. the top face of a Tetrahedron, which does not have a right angle
+    // 2. wedges and pyramids, whose faces do not lie on the reference simplex
+    //
+    // Deal with both by creating a locally orthogonal (but not necessarily
+    // orthonormal) coordinate system and testing if the projected point is in
+    // the triangle by expressing it as a convex combination of the vertices.
+    if (reference_cell == ReferenceCells::Triangle)
+      {
+        e0 = shifted_vertices[1] - shifted_vertices[0];
+        e1 = shifted_vertices[2] - shifted_vertices[0];
+        e1 -= (e0 * e1) * e0 / (e0.norm_square());
+      }
+    else
+      {
+        e0 = shifted_vertices[1] - shifted_vertices[0];
+        e1 = shifted_vertices[2] - shifted_vertices[0];
+      }
+    Assert(std::abs(e0 * e1) <= 1e-14, ExcInternalError());
+    // the quadrilaterals on pyramids and wedges don't necessarily have edge
+    // lengths of 1 so we cannot skip the denominator
+    const double   c0 = e0 * (shifted_p - vertex) / e0.norm_square();
+    const double   c1 = e1 * (shifted_p - vertex) / e1.norm_square();
+    const Point<3> projected_shifted_p = vertex + c0 * e0 + c1 * e1;
+
+    bool in_quad = false;
+    if (reference_cell == ReferenceCells::Triangle)
+      {
+        Tensor<2, 3> shifted_vertex_matrix;
+        for (unsigned int i = 0; i < 3; ++i)
+          shifted_vertex_matrix[i] = shifted_vertices[i];
+        const Tensor<1, 3> combination_coordinates =
+          invert(transpose(shifted_vertex_matrix)) * projected_shifted_p;
+        bool is_convex_combination = true;
+        for (unsigned int i = 0; i < 3; ++i)
+          is_convex_combination = is_convex_combination &&
+                                  (0.0 <= combination_coordinates[i]) &&
+                                  (combination_coordinates[i] <= 1.0);
+        in_quad = is_convex_combination;
+      }
+    else
+      in_quad = 0.0 <= c0 && c0 <= 1.0 && 0.0 <= c1 && c1 <= 1.0;
+
+    if (in_quad)
+      return shifted_p.distance_square(projected_shifted_p);
+    else
+      return std::numeric_limits<double>::max();
+  }
+} // namespace
+
+
+
+template <int dim>
+double
+ReferenceCell::distance_to_unit_cell_square(const Point<dim> &p) const
+{
+  AssertDimension(dim, get_dimension());
+
+  // Handle simple cases first:
+  Assert(dim != 0, ExcNotImplemented());
+  if (contains_point(p, 0.0))
+    return 0.0;
```

Comment:
1. Done. We need to do this regardless of how optimized this is since these distance functions don't have a concept of volume (they really just compute a distance to the boundary of the reference cell).